common/io/TimeoutManager.cpp

```diff
@@ -94,11 +94,7 @@ void TimeoutManager::CancelTimeout(timeout_id id) {
 
 TimeInterval TimeoutManager::ExecuteTimeouts(TimeStamp *now) {
   Event *e;
-  if (m_events.empty())
-    return TimeInterval();
```

Comment:
Sorry I don't follow @baltitenger , surely it's done the empty check here?

Or are you talking about some race condition where it gets emptied by another external process after this check?

Reply:
I only removed those lines because they were redundant. The actual problem is in the for loop's iteration expression: `e = m_events.top()` would run before the loop's test (`!m_events.empty()`), but this can call `.top()` on an empty queue. Since you didn't access `e` in that case, you didn't actually get a segfault, but it's still undefined behavior, so there is a check for it in glibc when compiled with `_GLIBCXX_ASSERTIONS`.

Reply:
> I only removed those lines because they were redundant.

But they weren't.

> The actual problem is in the for loop's iteration expression: `e = m_events.top()` would run before the loop's test (`!m_events.empty()`), but this can call `.top()` on an empty queue.

But it can't because it's covered by the fail early (which you've removed) which had already checked whether m_events was empty.

> Since you didn't access `e` in that case, you didn't actually get a segfault, but it's still undefined behavior, so there is a check for it in glibc when compiled with `_GLIBCXX_ASSERTIONS`.

Would you mind sharing the compiler error output please.

Reply:
There's no compile error, only a runtime crash (saying some assertion failed).
I'm not talking about the initial entry to the function, that was indeed protected by that check. However, on the next iteration of the loop,  the queue can be empty again.

Reply:
> There's no compile error, only a runtime crash on line 101 (saying some assertion failed).

Ah gotcha, in which case the next question is does the test catch it, if so it's probably worth adding to our CI so we can catch such issues automagically in future.

> I'm not talking about the initial entry to the function, that was indeed protected by that check. However, on the next iteration of the loop, the queue can be empty again.

Yes, I see now, if we had one event, it would pop it, then call m_events.top() which is undefined, although it wouldn't actually use that value afterwards.